### PR TITLE
gscreenshot: init at 3.4.0

### PR DESCRIPTION
**Pull request by davisrichard437**
###### Description of changes

`gscreenshot` (https://github.com/thenaterhood/gscreenshot) is a screenshot frontend (CLI and GUI) for a variety of screenshot backends. These changes provide a default, fully-functional X11 configuration. I was unsure how this might work, but I would love suggestions for how to idiomatically package other configurations (particularly the recommended Wayland default).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/gscreenshot/0001-Changing-paths-to-be-nix-compatible.patch
Comment (ambroisie) [inline pkgs/applications/graphics/gscreenshot/0001-Changing-paths-to-be-nix-compatible.patch:30]: Isn't this shebang line re-written by `buildPythonApplication` anyway?

Comment (davisrichard437) [reply pkgs/applications/graphics/gscreenshot/0001-Changing-paths-to-be-nix-compatible.patch]: I believe you are correct, that seems to be something I did in a broad swoop to get the `.desktop` file to work. I have just pushed the fix.

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (ambroisie) [inline pkgs/applications/graphics/gscreenshot/default.nix:75]: I believe this is unnecessary since `pname = "gscreenshot"`.

File: pkgs/top-level/all-packages.nix
Comment (ambroisie) [inline pkgs/top-level/all-packages.nix:29559]: A few things: 

1. Why do you have `xclip ? null` and `xdg-utils ? null` in the argument list if you explicitly add them to the argument list anyway.
2. Do you actually need `python310` or any `python3` would do?
3. This is not the usual nixpkgs way of using `callPackage`. You should avoid using explicit arguments: I think you should instead do the following:
    * accept `python3Packages` directly, and use that to provide the packages to `propagatedBuildInputs`.
    * remove all explicit arguments from your `callPackage` call

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (ambroisie) [inline pkgs/applications/graphics/gscreenshot/default.nix:30]: Are those necessary?

Comment (davisrichard437) [reply pkgs/applications/graphics/gscreenshot/default.nix:30]: They are, the build fails without either of them. The tests run when `doCheck` is enabled require a display and therefore fail. Regarding `strictDeps`, see the warning in [this section of the Nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-hooks); `wrapGAppsHook` doesn't function without it.

I borrowed these from other packages which I can no longer find, as the repository's search functionality no longer seems to return any results for code.

File: pkgs/top-level/all-packages.nix
Comment (davisrichard437) [reply pkgs/top-level/all-packages.nix:29559]: 1. I had those default to `null` because they used to be not entirely essential when I first packaged it, though it seems that has since changed.
2. Just tried it, `python3` works just fine.
3. What I was initially thinking with this was to package an X11 version and a Wayland version, which would maybe require explicit `callPackage` arguments. Would this be appropriate? Would there be a more idiomatic way?

I'll push the changes for these, but I'd still very much like to ask for someone's input on how to best do this.

Comment (ambroisie) [reply pkgs/top-level/all-packages.nix]: For an X11 and Wayland version, I think you could do e.g: the default version is X11, and add a `gscreenshot-wayland` which overrides some of the arguments. However I don't really know the usual way this is handled in nixpkgs, I would defer to somebody more knowledgeable.

**Comment by davisrichard437** [comment]
I have just adjusted the package to provide both X11 and Wayland support by default, providing `x11Support` and `waylandSupport` flags as arguments that I might hope a potential user might be able to override if they only want one or the other. Is this appropriate?

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (ambroisie) [inline pkgs/applications/graphics/gscreenshot/default.nix:23]: I don't think that's strictly necessary, but I'll let an actual nixpkgs committer chime in on whether to remove it.

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (davisrichard437) [reply pkgs/applications/graphics/gscreenshot/default.nix]: I didn't know those functions existed, thank you so much!

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (davisrichard437) [reply pkgs/applications/graphics/gscreenshot/default.nix:60]: I'll fix the patch for the time being, but do you know where I could find documentation on `makeDesktopItem`?

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (davisrichard437) [reply pkgs/applications/graphics/gscreenshot/default.nix:58]: I have my editor configured to run `nixpkgs-fmt` before saving, which caused the wrong indentation. Should I use a different formatter or report the error somewhere?

File: pkgs/applications/graphics/gscreenshot/default.nix
Comment (ambroisie) [inline pkgs/applications/graphics/gscreenshot/default.nix:52]: I'm not sure if that works? At least that's not how I usually see it written.

```suggestion
    ] ++ (with python3Packages; [
      pillow
      pygobject3
      setuptools
    ]) ++ lib.optional (!waylandSupport) python3Packages.xlib # not needed for wayland build
    ;
```

Comment (ambroisie) [reply pkgs/applications/graphics/gscreenshot/default.nix:52]: And about the logic itself, does that mean that `python3Packages` is only needed if _not_ building for wayland? I would imagine it is instead necessary if `x11Support` is turned on.

Comment (davisrichard437) [reply pkgs/applications/graphics/gscreenshot/default.nix]: There was a much simpler way to write it, I was being silly. Thanks for pointing it out!

**Comment by davisrichard437** [comment]
Just bumped the version to the new release, v3.4.0. I'm not really in any particular rush, but what all do I need to for this to be merged?

**Review by ambroisie — APPROVED** [review]

**Pull request merged**
